controller_manager/scripts/controller_group

```
@@ -0,0 +1,135 @@
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+#
+# Support controller groups so it is easy to spawn a set controllers and also switch
+# to any other group at run time.
+#
+# Author: Yong Li
+#
+import rospy
+from argparse import ArgumentParser
+from controller_manager import controller_manager_interface
+from controller_manager_msgs.srv import ListControllers, ListControllersRequest
+
+
+def _read_group(group):
+    ''''
+    Return 2 sets of controllers. The first set includes controllers required for the given group,
+    and the second set includes the controllers that should be stopped if the given group is chosen.
+    '''
+    groups = rospy.get_param('controller_groups', {})
+    desired_controllers = set(groups.get(group, ()))
+    if not desired_controllers:
+        rospy.logerr('There is no controller defined in group {}'.format(group))
+        return None, None
+    all_controllers = set()
+    for controllers in groups.values():
+        # Convert list to set to remove duplicates.
+        all_controllers = all_controllers | set(controllers)
+    return desired_controllers, all_controllers - desired_controllers
+
+
+def _get_loaded_controllers():
+    rospy.wait_for_service('controller_manager/list_controllers')
```

Shouldn't we pass a `timeout=N`, where I think `N` could be 5s or so. Otherwise this could get block here and the user won't know what's going on.

If you add the timeout you need a try - except block catching `ROSException`.

---

Or let it err out with the exception, although less pretty, that's not bad either.

---

Was following controller_manager_interface. I will add the a timeout and let it crash with the exception.